### PR TITLE
Refactoring latex writer

### PR DESCRIPTION
**Pull request by tk0miya**
Discussed at #2714

### CODE REVIEW

File: sphinx/writers/latex.py [hunk 1]
```diff
@@ -1360,25 +1376,10 @@ def depart_hlistcol(self, node):
...
-            res = '%sbp' % amount  # convert to 'bp'
-        elif unit == "px":
-            res = "%.3f\\sphinxpxdimen" % (float(amount))
-        elif unit == "%":
-            res = "%.3f\\linewidth" % (float(amount) / 100.0)
-        return res
+        try:
+            return rstdim_to_latexdim(width_str)
+        except ValueError:
+            self.builder.warn('dimension unit %s is invalid. Ignored.')
```
Comment (jfbu) [inline sphinx/writers/latex.py]: I am not good at Python, isn't a value for `%s` needed here ? (sorry if trivial question)
Besides, independently there is pre-existing check (I don't know from where), e.g.

```
/Users/foo/testallowpx/index.rst:13: ERROR: Error in "image" directive:
invalid option value: (option: "width"; value: '100.3.211')
not a positive measure of one of the following units:
"em" "ex" "px" "in" "cm" "mm" "pt" "pc" "%".

.. image:: sphinxwide.png
   :width: 100.3.211
```

This is issued only on first `make latex`. One must `rm -fr _build/doctrees` to regenerate error. But this is separate issue from this thread.

File: sphinx/writers/latex.py [hunk 2]
```diff
@@ -275,6 +275,22 @@ def escape_abbr(text):
     return re.sub('\.(?=\s|$)', '.\\@', text)
 
 
+def rstdim_to_latexdim(width_str):
+    """Convert `width_str` with rst length to LaTeX length."""
+    match = re.match('^([0-9.]+)\s*(\S*)$', width_str)
```
Comment (jfbu) [inline sphinx/writers/latex.py]: seems this regular expression allows `1.2.3` for example ? this would be wrong but is intercepted anyhow. (see other comment)

File: sphinx/writers/latex.py [hunk 3]
```diff
@@ -275,6 +275,22 @@ def escape_abbr(text):
...
+    if not match:
+        raise ValueError
+    res = width_str
+    amount, unit = match.groups()[:2]
+    if unit in ('', "px"):
+        res = "%s\\sphinxpxdimen" % amount
+    elif unit == 'pt':
+        res = '%sbp' % amount  # convert to 'bp'
```
Comment (jfbu) [inline sphinx/writers/latex.py:289]: memo: the reason for doing this is that Sphinx is multi-target and `pt` everywhere means `1/72in` except in TeX, where it means `1/72.27in`. The DocUtils latex writer does this conversion too. http://docutils.sourceforge.net/docs/user/latex.html#length-units

File: sphinx/writers/latex.py [hunk 4]
```diff
@@ -275,6 +275,22 @@ def escape_abbr(text):
...
 
+def rstdim_to_latexdim(width_str):
+    """Convert `width_str` with rst length to LaTeX length."""
+    match = re.match('^([0-9.]+)\s*(\S*)$', width_str)
+    if not match:
+        raise ValueError
+    res = width_str
+    amount, unit = match.groups()[:2]
+    if unit in ('', "px"):
+        res = "%s\\sphinxpxdimen" % amount
```
Comment (jfbu) [inline sphinx/writers/latex.py:287]: memo: the reason for using `px` as default unit is that this is what has been so far documented by Sphinx for all targets, when no unit is specified.
http://www.sphinx-doc.org/en/1.4.4/rest.html#images

So far LaTeX was considered as not supporting `px` but PDFLaTeX for at least ten years natively accepts it. To maintain multi-target coherence, using default `96px=1in` (or `4px=3bp`) seems the thing to do (thus Sphinx changes pdflatex default setting of `72px=1in`, or `1px=1bp` where `1bp=1/72in`)

**Comment by jfbu** [comment]
LGTM with nits. doc/rest.rst#Images should be updated (cf https://github.com/jfbu/sphinx/commit/cd0e19c3b4b11d3bde9c8ed338bb00b34dcf5cdd)

**Comment by jfbu** [comment]
Do you need the test on `w` (earlier code could return None)

```
        if 'width' in attrs:
            w = self.latex_image_length(attrs['width'])
            if w:
                include_graphics_options.append('width=%s' % w)
```

same on next code lines in latex.py

File: sphinx/writers/latex.py (hunk 1 above)
Comment (tk0miya) [inline sphinx/writers/latex.py]: Fixed.

The pre-existing check is executed by image directive on docutils.
https://sourceforge.net/p/docutils/code/HEAD/tree/trunk/docutils/docutils/parsers/rst/directives/__init__.py#l242
But this is a special behavior of the directive. so an image node made by 3rd party extension might  have invalid dimension units if the extension does not check.

File: sphinx/writers/latex.py (hunk 2 above)
Comment (tk0miya) [inline sphinx/writers/latex.py]: This is copied from docutils'.
https://sourceforge.net/p/docutils/code/HEAD/tree/trunk/docutils/docutils/parsers/rst/directives/__init__.py#l233

But this is not strict as you said. So I'll revert this.

**Comment by tk0miya** [comment]
`latex_image_length()` return none value when invalid width_str specified. So the after check is needed.
I noticed the codes for figure does not check that. Thank you for comment :-)

**Comment by jfbu** [comment]
LGTM !

File: sphinx/writers/latex.py [hunk 5]
```diff
@@ -275,6 +275,23 @@ def escape_abbr(text):
...
 
 
+def rstdim_to_latexdim(width_str):
+    """Convert `width_str` with rst length to LaTeX length."""
+    match = re.match('^(\d*\.?\d*)\s*(\S*)$', width_str)
+    if not match:
+        raise ValueError
+    res = width_str
+    amount, unit = match.groups()[:2]
+    float(amount)  # validate amount is float
```
Comment (jfbu) [inline sphinx/writers/latex.py:285]: just out of curiosity what could go wrong here if match is ok ?

Comment (tk0miya) [inline sphinx/writers/latex.py:285]: The regexp above is not strict to support multiple representations of amount. `1`, `1.0`, `1.` and `.5` are allowed. It matches even empty string or only dot (`.`).
This ensures the amount is valid float.

I know we can use more strict regexp instead. But it is a bit complicated and not readable.

**Comment by tk0miya** [comment]
Merged. Thank you for reviewing!
